bin/promotion.ml

```diff
@@ -50,12 +50,56 @@ module Apply = struct
     Cmd.info ~doc ~man "apply"
   ;;
 
+  let remote_promote files_to_promote =
+    let open Fiber.O in
+    let* connection = Rpc_common.establish_client_session ~wait:true in
+    Dune_rpc_impl.Client.client
+      connection
+      (Dune_rpc.Initialize.Request.create ~id:(Dune_rpc.Id.make (Sexp.Atom "promote")))
+      ~f:(fun client ->
+        Rpc_common.request_exn
+          client
+          (Dune_rpc_private.Decl.Request.witness
+             Dune_rpc_private.Procedures.Public.promote)
+          files_to_promote)
+  ;;
+
+  let promote_via_rpc_server files_to_promote =
+    let open Fiber.O in
+    remote_promote files_to_promote
+    >>| function
+    | Error (error : Dune_rpc_private.Response.Error.t) ->
+      Printf.eprintf
+        "Error: %s\n%!"
+        (Dyn.to_string (Dune_rpc_private.Response.Error.to_dyn error))
+    | Ok () ->
+      Console.print_user_message
+        (User_message.make [ Pp.text "Success" |> Pp.tag User_message.Style.Success ])
+  ;;
+
   let term =
     let+ builder = Common.Builder.term
     and+ files = Arg.(value & pos_all Cmdliner.Arg.file [] & info [] ~docv:"FILE") in
-    let common, _config = Common.init builder in
-    let files_to_promote = files_to_promote ~common files in
-    Diff_promotion.promote_files_registered_in_last_run files_to_promote
+    let common, config = Common.init builder in
+    match Dune_util.Global_lock.lock ~timeout:None with
```

Comment:
Tagging @Alizter to mention that this `lock` check is what was missing for synchronisation